core(renderer): improve the unknown timezone checks in util.js

petetnt opened the pull request:
<!-- Thank you for submitting a pull request! -->

**Summary**
<!-- What kind of change does this PR introduce? -->
<!-- Is this a bugfix, feature, refactoring, build related change, etc? -->

This PR is a triple-down fix for https://github.com/GoogleChrome/lighthouse/issues/1056, which was partly fixed in https://github.com/GoogleChrome/lighthouse/commit/2f9ad6ffa8348f699933b211d7eb255a1b698f4a and https://github.com/GoogleChrome/lighthouse/pull/1086.

<!-- Describe the need for this change -->

Turns out that `Intl.DateTimeFormat` itself too throws if the timeZone is invalid.


The issue itself was fixed in newer Chromium versions, as it affects eg. Node 10 but not Node 12
<!-- Link any documentation or information that would help understand this change -->

**Related Issues/PRs**
https://github.com/GoogleChrome/lighthouse/pull/1086
https://github.com/GoogleChrome/lighthouse/commit/2f9ad6ffa8348f699933b211d7eb255a1b6
https://github.com/GoogleChrome/lighthouse/issues/1056

petetnt commented:
Hi @patrickhulce!

A test would be nice but I need to think a bit how it can be done. The `timeZone` needs to be set during launch AFAIK (eg. with Node with `TZ=Etc/Unknown node`, no idea on how it works with a Chromium browser) so mocking it is pretty non-trivial 🤔

Easiest way to manually test this is running `TZ=Etc/Unknown node -e 'Intl.DateTimeFormat()'` with Node 10 and 12, where the latter one succeeds and the former fails with 

```
▶ TZ=Etc/Unknown node -e 'Intl.DateTimeFormat()'
'[eval]:1
Intl.DateTimeFormat()
     ^

RangeError: Unsupported time zone specified undefined
    at Object.DateTimeFormat (native)
    at [eval]:1:6
    at Script.runInThisContext (vm.js:96:20)
    at Object.runInThisContext (vm.js:303:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at evalScript (internal/bootstrap/node.js:587:27)
    at startup (internal/bootstrap/node.js:265:9)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)
```

Need to think about that for a bit :)

patrickhulce commented:
@petetnt we choose the timezone using `Util.numberDateLocale` so the test should be able to just temporarily override this value to something made up to exercise this block :)


```js
  let originalDateLocale;

  beforeEach(() => {
   originalDateLocale = Util.numberDateLocale;
  });

  afterEach(() => {
   Util.numberDateLocale = originalDateLocale;
  });

  it('recovers from invalid timezones', () => {
   Util.numberDateLocale = 'madeup locale';
    const timestamp = Util.formatDateTime('2017-04-28T23:07:51.189Z');
    assert.ok(
      timestamp.includes('Apr 27, 2017') ||
      timestamp.includes('Apr 28, 2017') ||
      timestamp.includes('Apr 29, 2017')
    );
  });
```

petetnt commented:
Turns out that the PR needs some adjustments: as per spec, the values that must be supported for the `timeZone` value is UTC

```
A conforming implementation is permitted to accept additional values, and then have implementation-defined behaviour instead of throwing a RangeError, for the following properties of options arguments:

The options property timeZone in the DateTimeFormat constructor, provided that the additional acceptable input values are case-insensitive matches of Zone or Link identifiers in the IANA time zone database and are canonicalized to Zone identifiers in the casing used in the database for the timeZone property of the object returned by DateTimeFormat.resolvedOptions, except that "Etc/GMT" shall be canonicalized to "UTC".
https://www.ecma-international.org/ecma-402/1.0/#sec-2
[[timeZone]] is either the String value "UTC" or undefined.
https://www.ecma-international.org/ecma-402/1.0/#sec-12.4

The ECMAScript Internationalization API defines a single time zone name, "UTC", which identifies the UTC time zone.

The Intl.DateTimeFormat constructor allows this time zone name; if the time zone is not specified, the host environment’s current time zone is used.
https://www.ecma-international.org/ecma-402/1.0/#sec-6.4

If tz is not undefined, then
a. Let tz be ToString(tz).
b. Convert tz to upper case as described in 6.1.
NOTE: If an implementation accepts additional time zone values, as permitted under certain conditions by the Conformance clause, different casing rules apply.
c. If tz is not "UTC", then throw a RangeError exception.
https://www.ecma-international.org/ecma-402/1.0/#sec-12.1.1.1
```

Will add the tests and update the implementation in the morning

paulirish commented:
@petetnt can you take a stab at those updates?

connorjclark commented:
Hi @petetnt, thanks for your help so far. We could reassign this PR to someone else to get it across the finish line - let us know if you'd like to finish this yourself.

petetnt commented:
Hey @connorjclark ! Sorry I completely forgot about this. I rebased the changes against the resent changes (time formatting is now under rendered/i18n, not in utils) and adjusted/simplified the checks against the spec (https://github.com/GoogleChrome/lighthouse/pull/9822#issuecomment-541929025) so that any throwing default value is set to `UTC`.

Also added a simple test that unknown timezones don't throw.

petetnt commented on lighthouse-core/report/html/renderer/i18n.js:
```diff
@@ -80,17 +80,20 @@ class I18n {
-    const tz = formatter.resolvedOptions().timeZone;
-    if (!tz || tz.toLowerCase() === 'etc/unknown') {
+    // and https://github.com/GoogleChrome/lighthouse/pull/9822
+    try {
+      (new Date()).toLocaleTimeString();
+    } catch (err) {
```
IIRC this was done in this peculiar way for a very specific reason but honestly cannot remember why exactly. (The comment should have reflect that, sigh...). 

Most likely just due to

1. Bikeshedding in https://github.com/webpack/webpack/issues/9818 where the similar issue was fixed 
2. not having to call `Intl.DateTimeFormat` twice
3. Avoiding any errornous values (mostly options.timeZone being set with a proper timeZone like CET when it only accepts UTC)

FWIW I don't have any strong feelings either way, as long as it works :)

patrickhulce replied:
Gotcha, thanks for the explanation! 

I'm not sure I understand the 3rd point though. Is that saying `formatter` can be created successfully with a  timeZone it doesn't support and only throw later on `.format` but that `(new Date()).toLocaleTimeString()` catches this ahead of time?

If that's true, just a comment here to that effect sounds great.

If I misunderstood then I vote my earlier suggestion :D I'm not all that convinced by the earlier points. If we were worried about excessive `Intl.DateTimeFormat` calls we would create the formatter when `_numberDateLocale` changes and not on every invocation of `formatDateTime` ;)

patrickhulce replied:
Friendly ping on this :) I'm happy to do the extra cleanup work here I just want to make sure I'm not discarding any of your valuable research you've already done!

petetnt replied:
Sorry for the late response, got back to work last week and this slipped through the cracks!

> If I misunderstood then I vote my earlier suggestion :D I'm not all that convinced by the earlier points. If we were worried about excessive Intl.DateTimeFormat calls we would create the formatter when _numberDateLocale changes and not on every invocation of formatDateTime ;)

Totally, you are right there. No sense to sacrifice readability for meaningless "DRY"ishness.

> I'm not sure I understand the 3rd point though. Is that saying formatter can be created successfully with a timeZone it doesn't support and only throw later on .format but that (new Date()).toLocaleTimeString() catches this ahead of time?

IIRC this was the reasoning in why the Webpack approach didn't work at the time, but it doesn't seem to affect this implementation so probably fine to go with your suggested changes :)

patrickhulce commented on lighthouse-core/report/html/renderer/i18n.js:
```diff
@@ -80,17 +80,20 @@ class I18n {
+    try {
+      (new Date()).toLocaleTimeString();
+    } catch (err) {
       options.timeZone = 'UTC';
```
```suggestion
      options.timeZone = 'UTC';
      formatter = new Intl.DateTimeFormat(this._numberDateLocale, options);
```

patrickhulce commented on lighthouse-core/report/html/renderer/i18n.js:
```diff
@@ -80,17 +80,20 @@ class I18n {
-      formatter = new Intl.DateTimeFormat(this._numberDateLocale, options);
     }
+
+    const formatter = new Intl.DateTimeFormat(this._numberDateLocale, options);
```
```suggestion
```

petetnt commented:
@patrickhulce 🥳  thanks for picking it back up!
